### PR TITLE
Update to arkworks 0.3.x

### PR DESCRIPTION
**Pull request by bhgomes**
- library builds but requires an old version of `rand_core` to make
  `x25519_dalek` happy, because of this, the tests won't build
- see `FIXME[check]` notes for things that have to be fixed by the time
  this PR merges
- see `TODO[check]` notes for things that may be fixed or may not, but
  the TODO message should be removed before PR merging

### CODE REVIEW

**Comment by bhgomes** [comment]
Now we get tests building, but one test fails because of a possible bug in arkworks.

**Comment by bhgomes** [comment]
v0.3.0 Tag: https://github.com/arkworks-rs/crypto-primitives/blob/5b0d4ba293aa91b65326eb934e4b2322fc06a4f2/src/merkle_tree/mod.rs#L158-L241
Development Branch: https://github.com/arkworks-rs/crypto-primitives/blob/cde6a5193d278acd911ee79349a021d1c39149aa/src/merkle_tree/mod.rs#L253-L324

Both panic on size 1 trees because of an `Option::unwrap`. The `level_indices` variable is a length zero vector in the case of a size 1 tree, so popping from it an unwrapping is a panic.

This is my understanding of whats going on at least.

**Comment by stechu** [comment]
> v0.3.0 Tag: https://github.com/arkworks-rs/crypto-primitives/blob/5b0d4ba293aa91b65326eb934e4b2322fc06a4f2/src/merkle_tree/mod.rs#L158-L241
> Development Branch: https://github.com/arkworks-rs/crypto-primitives/blob/cde6a5193d278acd911ee79349a021d1c39149aa/src/merkle_tree/mod.rs#L253-L324
> 
> Both panic on size 1 trees because of an `Option::unwrap`. The `level_indices` variable is a length zero vector in the case of a size 1 tree, so popping from it and unwrapping is a panic.
> 
> This is my understanding of what's going on at least.

Indeed looks like their bug, can you file a bug report to them?

**Comment by stechu** [comment]
@bhgomes  after reading the arkworks code, I think they assume that the `tree_height` is at least 2. I agree that they should make this assumption more explicit, but I don't think this is a bad assumption.

**Comment by bhgomes** [comment]
I think it's still an error for `leaves.len() == 1` to give a naked panic. Since they have an `assert!` for powers of two they should at least have one for this case too. I realize that the `Path` data structure would not work as currently designed if you kept the length 1 assumption. I still think a Merkle Tree with one leaf is a valid tree but if we don't need the this case then that's ok as long as we can prove we won't get there.

**Comment by stechu** [comment]
agreed! seems to be a small fix.

**Comment by bhgomes** [comment]
We should close #22 since its now outdated.

File: src/checksum.rs [hunk 1]
```diff
@@ -17,82 +17,79 @@
...
-		let final_digest = hasher_two.finalize();
-		let mut res = [0u8; 4];
-		res.copy_from_slice(&final_digest.as_slice()[0..4]);
-		Ok(res)
+		Ok(checksum(&checksum(self))[0..4].try_into().unwrap())
 	}
 }
+
+#[inline]
+fn checksum(input: &[u8]) -> [u8; 32] {
```
Comment (stechu) [inline src/checksum.rs:88]: Thanks for the refactoring!

File: src/ecies.rs [hunk 2]
```diff
@@ -107,23 +120,23 @@ impl Ecies for MantaCrypto {
...
+
 		Ok(MantaEciesCiphertext {
-			encrypted_msg: ciphertext_bytes,
+			encrypted_msg,
 			ephemeral_pk: ephemeral_pk_bytes,
 		})
 	}
 
-	/// Decrypt a ciphertext with a secret key.
+	/// Decrypts the ciphertext using the given secret key.
```
Comment (stechu) [inline src/ecies.rs:139]: should be Decrypt here since this is an imperative sentence?

File: src/constants.rs [hunk 3]
```diff
@@ -14,12 +14,55 @@
...
+/// Returns the fixed commitment parameters.
+#[inline]
+pub fn commitment_parameters() -> CommitmentParam {
+	try_commitment_parameters().expect("Failed to deserialize `COMMIT_PARAM.data`")
+}
+
+/// Tries to return the fixed leaf hash parameters.
+#[inline]
+pub fn try_leaf_parameters() -> Result<LeafHashParam, MantaError> {
+	let mut bytes = HASH_PARAM.data;
```
Comment (stechu) [inline src/constants.rs:36]: So the `HASH_PARAM` itself doesn't change due to we stick to the same hash function?

**Review by stechu — COMMENTED** [review]
LGTM in general! Leave a few questions here.

**Pull request merged**